### PR TITLE
fix(components): [time-picker] avoid triggering `change` twice when clearing

### LINKED ISSUE #24826: [Component] [date-picker] 日期选择器清空时触发两次 `change` 的完整机制分析
Bug Type: **`Component`**

## Environment

- Vue Version: `3.3.4`
- Element Plus Version: `2.14.5`
- Browser / OS: `Google Chrome 152.0.7977.66 / windows`
- Build Tool: `Vite`

### Related Component

- `el-date-picker`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtZGF0ZS1waWNrZXJcbiAgICAgIDptb2RlbC12YWx1ZT1cInRvUGlja2VyVmFsdWUoZGF0ZSlcIlxuICAgICAgdHlwZT1cImRhdGVcIlxuICAgICAgY2xlYXJhYmxlXG4gICAgICBAdXBkYXRlOm1vZGVsVmFsdWU9XCJoYW5kbGVVcGRhdGVcIlxuICAgICAgQGNoYW5nZT1cImhhbmRsZUNoYW5nZVwiXG4gIC8+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IHNldHVwPlxuaW1wb3J0IHsgcmVmIH0gZnJvbSAndnVlJ1xuXG5jb25zdCBkYXRlID0gcmVmKCcnKVxuXG4vLyDmiornqbrlgLznu5/kuIDovazmiJDnqbrlrZfnrKbkuLLvvIzmlrnkvr/lrZjlgqhcbmZ1bmN0aW9uIHRvUGlja2VyVmFsdWUodmFsdWUpIHtcbiAgaWYgKCF2YWx1ZSkgcmV0dXJuICcnXG4gIHJldHVybiB2YWx1ZVxufVxuXG5mdW5jdGlvbiBoYW5kbGVVcGRhdGUodmFsKSB7XG4gIGRhdGUudmFsdWUgPSB2YWwgPz8gJydcbn1cblxuZnVuY3Rpb24gaGFuZGxlQ2hhbmdlKHZhbCkge1xuICBjb25zb2xlLmxvZygnW2NoYW5nZV0nLCB2YWwpXG59XG48L3NjcmlwdD4iLCJlbGVtZW50LXBsdXMuanMiOiJpbXBvcnQgRWxlbWVudFBsdXMgZnJvbSAnZWxlbWVudC1wbHVzJ1xuaW1wb3J0IHsgZ2V0Q3VycmVudEluc3RhbmNlIH0gZnJvbSAndnVlJ1xuXG5sZXQgaW5zdGFsbGVkID0gZmFsc2VcbmF3YWl0IGxvYWRTdHlsZSgpXG5cbmV4cG9ydCBmdW5jdGlvbiBzZXR1cEVsZW1lbnRQbHVzKCkge1xuICBpZiAoaW5zdGFsbGVkKSByZXR1cm5cbiAgY29uc3QgaW5zdGFuY2UgPSBnZXRDdXJyZW50SW5zdGFuY2UoKVxuICBpbnN0YW5jZS5hcHBDb250ZXh0LmFwcC51c2UoRWxlbWVudFBsdXMpXG4gIGluc3RhbGxlZCA9IHRydWVcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIGxvYWRTdHlsZSgpIHtcbiAgY29uc3Qgc3R5bGVzID0gWydodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvZGlzdC9pbmRleC5jc3MnLCAnaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L3RoZW1lLWNoYWxrL2RhcmsvY3NzLXZhcnMuY3NzJ10ubWFwKChzdHlsZSkgPT4ge1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgICAgbGluay5ocmVmID0gc3R5bGVcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgICB9KVxuICB9KVxuICByZXR1cm4gUHJvbWlzZS5hbGxTZXR0bGVkKHN0eWxlcylcbn1cbiIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9kaXN0L2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcImh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9cIixcbiAgICBcIkBlbGVtZW50LXBsdXMvaWNvbnMtdnVlXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AZWxlbWVudC1wbHVzL2ljb25zLXZ1ZUAyL2Rpc3QvaW5kZXgubWluLmpzXCJcbiAgfSxcbiAgXCJzY29wZXNcIjoge31cbn0iLCJfbyI6e319)

### Steps to reproduce

# 日期选择器清空时触发两次 `change` 的完整机制分析

## 一、问题现象

在使用 Element Plus 的 `el-date-picker` 时，如果父组件在更新绑定值时将空值 `null` 转换为空字符串 `''`：

```vue
<template>
  <el-date-picker
    :model-value="toPickerValue(date)"
    type="date"
    clearable
    @update:modelValue="handleUpdate"
    @change="handleChange"
  />
</template>

<script setup>
import { ref } from 'vue'

const date = ref('')

function toPickerValue(value) {
  if (!value) return ''
  return value
}

function handleUpdate(val) {
  date.value = val ?? ''
}

function handleChange(val) {
  console.log('[change]', val)
}
</script>
```

操作：先选择一个日期，然后点击清空按钮（×）。

控制台输出：

```
[change] null
[change] ''
```

**触发条件**：此现象仅在点击清空按钮时**输入框未聚焦**才会出现；若输入框已聚焦，则只会触发一次 `change`，值为 `null`。

---

## 二、事件流全貌

点击清空按钮时，如果输入框未聚焦，整个事件流如下：

```
click 事件捕获阶段
    ↓
ElInput 的 handleClick（捕获监听器）
    → 输入框未聚焦，执行 input.focus()
    → focus 事件触发 afterFocus
    → pickerVisible = true（面板打开）
    ↓
浏览器微任务检查点（watch 打开分支执行）
    ↓
click 事件目标/冒泡阶段
    ↓
清空按钮的 onClearIconClick
    → emitInput(null)（父组件更新绑定值）
    → emitChange(null, true)（第一次 change）
    → onHide() → pickerVisible = false（面板关闭）
    ↓
浏览器微任务检查点（watch 关闭分支执行）
    → nextTick 中 emitChange(props.modelValue)
    → 比较发现值不相等
    → 第二次 change
```

核心问题：**点击清空按钮时，组件错误地在聚焦输入框后打开了面板。** 聚焦输入框本身是合理的，但打开面板是不应该发生的，这个多余的面板打开与随后关闭，最终引发了第二次 `change`。

下面逐步拆解每一个环节。

---

### 步骤 1：捕获阶段，`ElInput` 的聚焦与面板误开

`click` 事件从 `window` 向下传播。经过 `ElInput` 容器时，触发**捕获阶段**绑定的 `handleClick`：

```js
const handleClick = (event) => {
  if (
    disabled.value ||
    isFocusable(event.target) ||
    (wrapperRef.value?.contains(document.activeElement) &&
      wrapperRef.value !== document.activeElement)
  )
    return

  target.value?.focus()
}
```

输入框未聚焦时，条件不满足，执行 `input.focus()`。  
`focus()` 触发 `focus` 事件，同步调用 `handleFocus` → `afterFocus`：

```js
function afterFocus() {
  if (!props.automaticDropdown) return
  pickerVisible.value = true   // 面板打开
}
```

此时面板被打开。**这一步是 bug 的起点：聚焦输入框是正常行为，但打开面板是不应该的。**

> **注意**：如果输入框已经聚焦，`handleClick` 会因焦点已在内部而直接 `return`，面板不会打开，后续不会出现第二次 `change`。这解释了为什么触发条件是“输入框未聚焦”。

---

### 步骤 2：微任务检查点，`watch` 触发打开分支

捕获监听器执行完毕，浏览器执行**微任务检查点**。

Vue 的 `watch(pickerVisible)` 打开分支被触发：

```js
watch(pickerVisible, (val) => {
  if (!val) {
    // 关闭分支，稍后分析
  } else {
    nextTick(() => {
      if (val) {
        valueOnOpen.value = props.modelValue  // 记录旧日期
      }
    })
  }
})
```

`valueOnOpen` 被记录为清空前的旧日期。

---

### 步骤 3：目标/冒泡阶段，清空按钮执行清空逻辑

事件继续传播到清空图标，触发 `onClearIconClick`：

```js
const onClearIconClick = (event) => {
  event.stopPropagation()
  emitInput(null)                    // 发出 update:modelValue
  emitChange(null, true)             // 第一次 change
  onHide()                           // 关闭面板
}
```

依次发生三件事：

#### 3.1 `emitInput(null)` → 父组件更新绑定值

父组件 `handleUpdate` 收到 `null`，将其转换为 `''`：

```js
function handleUpdate(val) {
  date.value = val ?? ''   // null 被转成 ''
}
```

`props.modelValue` 变为 `''`（Vue 响应式更新异步完成，但会在当前微任务阶段处理）。

#### 3.2 `emitChange(null, true)` → 第一次 `change`

`emitChange` 源码：

```js
const emitChange = (val, isClear) => {
  if (isClear || !valueEquals(val, valueOnOpen.value)) {
    emit(CHANGE_EVENT, val);
    isClear && (valueOnOpen.value = val);   // 清空时同步更新 valueOnOpen
  }
};
```

此时 `emitChange` 接收到的参数：

- `val` = `null`
- `isClear` = `true`
- `valueOnOpen.value` = 旧日期（面板打开时记录的）

因为 `isClear` 为 `true`，条件直接成立：

1. 触发第一次 `change`，值为 `null`。
2. 执行 `valueOnOpen.value = null`，将 `valueOnOpen` 从旧日期更新为 `null`。

#### 3.3 `onHide()` → 关闭面板

```js
pickerVisible.value = false
```

面板关闭。

---

### 步骤 4：微任务检查点，`watch` 触发关闭分支

目标/冒泡监听器执行完毕，浏览器再次执行**微任务检查点**。

此时状态：

- `props.modelValue` = `''`（父组件已更新）
- `valueOnOpen` = `null`（清空逻辑已同步设置）
- `pickerVisible` 从 `true` 变为 `false`，触发 `watch` 关闭分支：

```js
watch(pickerVisible, (val) => {
  if (!val) {
    userInput.value = null;
    nextTick(() => {
      emitChange(props.modelValue);   // 第二次 change 的源头
    });
  }
})
```

在 `nextTick` 中执行 `emitChange(props.modelValue)`。

此时 `emitChange` 接收到的参数：

- `val` = `''`（父组件转换后的 `props.modelValue`）
- `isClear` = `undefined`（未传）
- `valueOnOpen.value` = `null`（清空时已设置）

内部比较：

```js
if (isClear || !valueEquals(val, valueOnOpen.value)) {
  emit(CHANGE_EVENT, val)
}
```

`valueEquals('', null)` 比较不等，条件成立，触发第二次 `change`，值为 `''`。

---

## 四、关键机制：为什么面板的打开和关闭没有被合并？

一个自然会产生的疑问：`pickerVisible` 在同一个 `click` 事件传播过程中先被设为 `true` 又被设为 `false`，最终值等于初始值。按道理 Vue 的 `watch` 不应该触发，但实际触发了。为什么？

### 实验验证

最小化 demo：

```html
<div id="outer"><button id="btn">点我测试</button></div>
<script>
  let _value = false;
  let dirty = false;
  let lastWritten = false;

  const state = {
    get value() { return _value; },
    set value(v) {
      _value = v;
      lastWritten = v;
      if (!dirty) {
        dirty = true;
        Promise.resolve().then(() => {
          console.log('watch 触发，最终值:', lastWritten);
          dirty = false;
        });
      }
    }
  };

  const outer = document.getElementById('outer');

  outer.addEventListener('click', () => {
    console.log('listener 1: set true');
    state.value = true;
  });

  outer.addEventListener('click', () => {
    console.log('listener 2: set false');
    state.value = false;
  });
</script>
```

输出：

```
listener 1: set true
watch 触发，最终值: true
listener 2: set false
watch 触发，最终值: false
```

### 原因：浏览器在每个事件监听器之间执行微任务检查点

浏览器事件传播过程中，**每执行完一个事件监听器，就会清空微任务队列**。

因此：

- `handleClick`（捕获监听器）执行完毕 → 微任务检查点 → `watch` 打开分支被触发。
- 然后才执行 `onClearIconClick`（目标监听器）→ 微任务检查点 → `watch` 关闭分支被触发。
- 两次 `pickerVisible` 的修改发生在**不同的微任务周期**中，Vue 无法将它们合并。

这就是为什么面板的“打开→关闭”会导致 `watch` 分别触发两次，而不是被合并为一次。

---

### 本质

**点击清空按钮时，输入框获得焦点是合理行为，但组件不应因此打开面板。**

当前实现中，`afterFocus` 无条件地在聚焦后打开面板，导致清空按钮的点击意外打开了面板。这个多余的面板打开与随后清空逻辑中的关闭，使 `pickerVisible` 经历了 `false → true → false`。由于浏览器在每个事件监听器之间执行微任务检查点，Vue 无法合并这些修改，`watch` 分别触发了打开和关闭分支。关闭分支的额外检查最终导致第二次 `change`。

### 修复方向

聚焦输入框本身不应被禁止，**修复的关键是在 `afterFocus` 中判断聚焦来源，如果聚焦是由清空按钮点击引起的，则不打开面板**。

例如，可以在 `onClearIconClick` 中设置一个标志位，并在 `afterFocus` 中检查：

```js
let isClearing = false;

const onClearIconClick = (event) => {
  isClearing = true;
  // ... 清空逻辑
  onHide();
  isClearing = false;
};

function afterFocus() {
  if (!props.automaticDropdown) return;
  if (isClearing) return;   // 清空引发的聚焦不打开面板
  pickerVisible.value = true;
}
```

---

## 六、业务侧规避方案（临时）

由于组件修复需要等待官方更新，业务上应遵循组件空值语义，**不要在 `update:modelValue` 中转换空值类型**：

```vue
<template>
  <el-date-picker v-model="date" type="date" clearable @change="handleChange" />
</template>

<script setup>
import { ref } from 'vue'

const date = ref(null)  // 初始为 null

function handleChange(val) {
  console.log('[change]', val)  // 清空时只有一次 change，值为 null
}

function submit() {
  const payload = {
    date: date.value ?? ''  // 提交时再转成空字符串
  }
  // 发送请求
}
</script>
```

这样清空后 `props.modelValue` 保持 `null`，`valueOnOpen` 也是 `null`，面板关闭时比较相等，不会触发第二次 `change`。

---

## 七、总结

1. **第一次 `change`**：清空按钮自身触发，值为 `null`，同时将 `valueOnOpen` 同步设置为 `null`。
2. **第二次 `change`**：仅当输入框未聚焦时发生。聚焦后 `afterFocus` 错误地打开了面板，随后清空逻辑关闭面板。由于浏览器在每个事件监听器之间执行微任务检查点，Vue 无法合并 `pickerVisible` 的两次变化。关闭分支中的额外检查发现 `props.modelValue`（父组件转换为 `''`）与 `valueOnOpen`（`null`）不相等，触发第二次 `change`。
3. **根本缺陷**：清空按钮点击导致的聚焦不应打开面板。
4. **规避方法**：保持绑定值为 `null`，格式化延后到提交阶段。

### What is Expected?

只触发一次 change，值为 null。

### What is actually happening?

触发两次 change：
第一次值为 null（由清空按钮自身触发）
第二次值为 ''（由面板关闭后的额外检查触发）

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by patienry**
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

fix #24826

In fact, clearing the value while the component is focused also causes an issue.

The reason is that the parent component sets `model-value` in `@update:modelValue`. When the dropdown closes, `emitChange(props.modelValue)` is called, but `props.modelValue` has already been updated at that point, causing the `change` event to be triggered again.

[demo](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtdGltZS1waWNrZXJcbiAgICA6bW9kZWwtdmFsdWU9XCJkYXRlXCJcbiAgICBjbGVhcmFibGVcbiAgICBAdXBkYXRlOm1vZGVsVmFsdWU9XCJoYW5kbGVVcGRhdGVcIlxuICAgIEBjaGFuZ2U9XCJoYW5kbGVDaGFuZ2VcIlxuICAvPlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgZGF0ZSA9IHJlZigpXG5cbmZ1bmN0aW9uIGhhbmRsZVVwZGF0ZSh2YWwpIHtcbiAgZGF0ZS52YWx1ZSA9IHZhbCA/PyAnJ1xufVxuXG5mdW5jdGlvbiBoYW5kbGVDaGFuZ2UodmFsKSB7XG4gIGNvbnNvbGUubG9nKCdbY2hhbmdlXScsIHZhbClcbn1cbjwvc2NyaXB0PlxuIiwiZWxlbWVudC1wbHVzLmpzIjoiaW1wb3J0IEVsZW1lbnRQbHVzIGZyb20gJ2VsZW1lbnQtcGx1cydcbmltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcblxubGV0IGluc3RhbGxlZCA9IGZhbHNlXG5hd2FpdCBsb2FkU3R5bGUoKVxuXG5leHBvcnQgZnVuY3Rpb24gc2V0dXBFbGVtZW50UGx1cygpIHtcbiAgaWYgKGluc3RhbGxlZCkgcmV0dXJuXG4gIGNvbnN0IGluc3RhbmNlID0gZ2V0Q3VycmVudEluc3RhbmNlKClcbiAgaW5zdGFuY2UuYXBwQ29udGV4dC5hcHAudXNlKEVsZW1lbnRQbHVzKVxuICBpbnN0YWxsZWQgPSB0cnVlXG59XG5cbmV4cG9ydCBmdW5jdGlvbiBsb2FkU3R5bGUoKSB7XG4gIGNvbnN0IHN0eWxlcyA9IFsnaHR0cHM6Ly9yYXcuZXNtLnNoL3ByL2VsZW1lbnQtcGx1c0AyNDg0MS9kaXN0L2luZGV4LmNzcycsICdodHRwczovL3Jhdy5lc20uc2gvcHIvZWxlbWVudC1wbHVzQDI0ODQxL3RoZW1lLWNoYWxrL2RhcmsvY3NzLXZhcnMuY3NzJ10ubWFwKChzdHlsZSkgPT4ge1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgICAgbGluay5ocmVmID0gc3R5bGVcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgICB9KVxuICB9KVxuICByZXR1cm4gUHJvbWlzZS5hbGxTZXR0bGVkKHN0eWxlcylcbn1cbiIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vcmF3LmVzbS5zaC9wci9lbGVtZW50LXBsdXNAMjQ4NDEvZGlzdC9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJodHRwczovL3Jhdy5lc20uc2gvcHIvZWxlbWVudC1wbHVzQDI0ODQxL1wiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsIl9vIjp7InNob3dIaWRkZW4iOnRydWUsInN0eWxlU291cmNlIjoiaHR0cHM6Ly9yYXcuZXNtLnNoL3ByL2VsZW1lbnQtcGx1c0AyNDg0MS9kaXN0L2luZGV4LmNzcyJ9fQ==)

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed duplicate change events when clearing date or time picker values.
  - Clearing a value now emits a single change event with a `null` value.
  - Prevented the picker panel from opening when clearing a value while the picker is unfocused.
  - Improved handling when cleared values are normalized to an empty value.
  - Fixed an issue where canceling a clear action could prevent the date picker from reopening when the input regains focus.

- **Tests**
  - Added regression coverage for date and time picker clearing and focus behavior.

### CODE REVIEW

File: packages/components/time-picker/__tests__/time-picker.test.tsx [hunk 1]
```diff
@@ -585,6 +585,25 @@ describe('TimePicker', () => {
...
   })
+
+  it('should only trigger change once when clearing', async () => {
+    const date = ref(new Date())
+    const changeHandler = vi.fn()
+    const wrapper = mount(() => (
+      <TimePicker
+        modelValue={date.value}
+        clearable
+        onUpdate:modelValue={(val) => (date.value = val ?? '')}
```
Comment (patienry) [inline packages/components/time-picker/__tests__/time-picker.test.tsx]: This usage is relatively uncommon, so adding a `shouldSkipChange` just to fix it may not be a good approach. The issue in #24826 can be avoided by using `value-on-clear`. 🤔

Comment (patienry) [inline packages/components/time-picker/__tests__/time-picker.test.tsx:596]: Developers can use `value-on-clear` to avoid handling it this way, so perhaps we don't need to add `shouldSkipChange` to fix this. 🤔

Comment (btea) [reply packages/components/time-picker/__tests__/time-picker.test.tsx:596]: Yes, that's true. However, this kind of edge case might need to be addressed.

I made a change and submitted it. Please take a look and see if it's reasonable.

File: packages/components/time-picker/src/common/picker.vue [hunk 2]
```diff
@@ -452,8 +467,13 @@ const showClearBtn = computed(
     (hovering.value || isFocused.value)
 )
 
+const onClearIconMousedown = () => {
+  isClearIconMousedown = true
+}
```
[the remark this answers is not in the document]
Comment (btea) [reply packages/components/time-picker/src/common/picker.vue:473]: The latest submissions have tried to avoid this problem.

File: packages/components/time-picker/__tests__/time-picker.test.tsx (hunk 1 above)
Comment (patienry) [reply packages/components/time-picker/__tests__/time-picker.test.tsx:596]: Thanks!

File: packages/components/time-picker/src/common/picker.vue [hunk 3]
```diff
@@ -348,7 +355,16 @@ const emitChange = (
...
-  if (isClear || !valueEquals(val, valueOnOpen.value)) {
+  if (
+    isClear ||
+    !(
+      valueEquals(val, valueOnOpen.value) ||
+      // the parent may normalize the cleared value to another empty
+      // representation, e.g. `null` to '' (#24826)
+      (emptyValues.isEmptyValue(val) &&
+        emptyValues.isEmptyValue(valueOnOpen.value))
+    )
```
Comment (patienry) [inline packages/components/time-picker/src/common/picker.vue:366]: It looks like the new validation here can already prevent `change` from being triggered twice. Although `isClearIconMousedown` can prevent `pickerVisible` from being updated incorrectly, it also introduces additional code. I'm not sure whether it's worth making so many changes for an edge case. 😵‍💫

Comment (btea) [reply packages/components/time-picker/src/common/picker.vue:366]: Yes, but we have to solve this problem, and I currently don't have any other better solutions.
